src/main.rs

```diff
@@ -1,60 +1,254 @@
+    fn from_env() -> Self {
         dotenv::dotenv().ok();
+
         let rpc_url: String = env::var("BITCOIN_RPC_URL").expect("BITCOIN_RPC_URL must be set");
         let rpc_user: String = env::var("BITCOIN_RPC_USER").expect("BITCOIN_RPC_USER must be set");
         let rpc_password: String =
             env::var("BITCOIN_RPC_PASSWORD").expect("BITCOIN_RPC_PASSWORD must be set");
-        Client::new(&rpc_url, Auth::UserPass(rpc_user, rpc_password)).unwrap()
+
+        RpcCredentials {
+            rpc_url,
+            rpc_user,
+            rpc_password,
+        }
+    }
+}
+
+struct RpcCookieCredentials {
+    url: String,
+    pathbuf: PathBuf,
+}
+
+impl LoadCredentials for RpcCookieCredentials {
+    fn from_env() -> Self {
+        dotenv::dotenv().ok();
+
+        let cookie_path = env::var("COOKIE_FILE").expect("Cookie file not set");
+        let url = env::var("BITCOIN_RPC_URL").expect("BITCOIN_RPC_URL not set");
+
+        RpcCookieCredentials {
+            pathbuf: PathBuf::from_str(&cookie_path).expect("Invalid cookie path"),
+            url,
+        }
+    }
+}
+
+lazy_static! {
+    static ref RPC_CLIENT: Client = {
+        // const TIMEOUT_UTXO_SET_SCANS: time::Duration = time::Duration::from_secs(60 * 8); // 8 minutes
+        // let RpcCredentials {
+        //     rpc_url,
+        //     rpc_user,
+        //     rpc_password,
+        // } = RpcCredentials::from_env();
+        // let custom_timeout_transport = jsonrpc::simple_http::Builder::new()
+        //     .url(&rpc_url)
+        //     .expect("invalid rpc url")
+        //     .auth(rpc_user, Some(rpc_password))
+        //     .timeout(TIMEOUT_UTXO_SET_SCANS)
+        //     .build();
+        // let custom_timeout_rpc_client =
+        //     jsonrpc::client::Client::with_transport(custom_timeout_transport);
+        // Client::from_jsonrpc(custom_timeout_rpc_client)
+
+
+        let creds = RpcCookieCredentials::from_env();
+        match Client::new(&creds.url, Auth::CookieFile(creds.pathbuf)) {
+            Ok(client) => client,
+            Err(err) => {
+                eprintln!("Error connecting to client: {:?}", err);
+                panic!()
+            }
+        }
     };
 }
 
-// static client: Client = Client::new("url", Auth::UserPass("user".to_owned(), "password".to_owned())).unwrap();
+fn get_block_by_height(block_height: u64) -> Result<block::Block> {
+    let rpc = &*RPC_CLIENT;
+    let block = rpc.get_block_hash(block_height)?;
+    let block = rpc.get_block(&block)?;
+    Ok(block)
+}
+
+fn get_block_time(block_height: u64) -> Result<Duration> {
+    let block = get_block_by_height(block_height)?;
+    Ok(Duration::seconds(block.header.time as i64))
+}
+
+/**
+ * Attempts to find average block time of recent blocks
+ */
+fn avg_time_to_mine(block_height: u64) -> Result<Duration> {
+    // infeasable to use this because querying takes too long
+    // should probably caluclate based on difficulty
```

Comment:
There might be a more efficient way to do this, but you're in the right general direction imo. Calculating based on difficulty won't give you a real-time measurement.

Maybe you only need the most recent block, the first block since the last adjustment and the total number of blocks mined so far to get an average.

Reply:
Oh sweet yea, I just pushed an update that works much better now, thanks! Hard to test exactly because I'm still syncing mainnet on my computer so it says should have been mined 2,260 days ago, but I'm at block height ~522k so that lines up.

Yea you're right difficulty wouldn't do it exactly. Maybe if I new an estimate of network hashrate then I could do it, but I think network hashrate is estimated based on difficulty and recent block times